lib/upgrader.js

```diff
@@ -0,0 +1,97 @@
+module.exports = class Upgrader {
+  constructor (session, hooks = {}) {
+    this.session = session
+    this.active = null
+    this.preupgrade = hooks.preupgrade || null
+  }
+
+  cancel () {
+    if (this.active) {
+      this.active.cancel()
+      this.active = null
+    }
+  }
+
+  onupgrade (request) {
+    if (this.active) this.active.cancel()
```

Comment:
This ensures that only 1 upgrade is active at a time by cancelling previous upgrades as new ones are applied. Not 100% sure if that's what we're after though.

Reply:
either this, or it "debounces", ie reruns the hook